README.md
## Mock Server
Simple mock server for REST and XML-RPC API

It can mock GET, POST, PUT, PATCH, DELETE and some more rarely used HTTP methods. 

Build status at [Travis CI](http://travis-ci.org/): [![Build Status](https://travis-ci.org/tomashanacek/mock-server.png?branch=master)](https://travis-ci.org/tomashanacek/mock-server)

## DEMO

http://mockapi.org/__manage

## Installation

Install with pip:

    $ pip install mock-server

Or install with easy_install

    $ easy_install mock-server

And after run

    $ mock-server --dir=/path/to/api
    
It will be listening on port 8888 and wait for your HTTP requests.

## Overview

Mocking service is as simple as creating directory and few files within it just like that:

    response content format: %METHOD%_%STATUS%.%FORMAT%
    response headers format: %METHOD%_H_%STATUS%.%FORMAT%

    root_dir/
        GET_200.json         # response content for GET /
        GET_H_200.json       # headers for GET /
        user/
            DELETE_404.xml   # response content for DELETE /user.xml?__statusCode=404
            POST_200.json    # response content for POST /user
            POST_H_200.json  # headers for POST /user
            

Or you can use web interface

    http://mockapi.org/__manage/create


Mocked GET /user/tom 

    $ curl -v -X GET http://mockapi.org/user/tom
    
    > GET /user/tom HTTP/1.1
    > Host: mockapi.org
    > Accept: */*
    > 
    < HTTP/1.1 200 OK
    < Access-Control-Allow-Origin: *
    < Content-Type: application/json; charset=utf-8
    < Content-Length: 64
    < Server: TornadoServer/2.4.1
    < 
    {
        "name": "Tom",
        "surname": "Smith",
        "age": 22
    }


## Bug report
If you have any trouble, report bug at GitHub Issue https://github.com/tomashanacek/mock-server/issues


CHANGES.rst
History
=======

0.3.9
----------------
- fix unicode 500 error
- fix grammar when API does not exist

Many thanks to @bopo and @krmaxwell

0.3.8
----------------

- fix http util when returning HTTPHeaders instance
- fix Tornado version
- Python 2.7 and 3.4 compatibility

Many thanks to @eduardosan and @winkidney

0.3.7
----------------

- support for custom response provider

0.3.6
----------------

- support for non-standard status codes

0.3.5
----------------

- added num_processes command line option for possibility to start multiple worker processes

0.3.4
----------------

- edit rpc method fix

0.3.3
----------------

- create resource quotation character fix

0.3.2
----------------

- todo escaping fix

0.3.1
----------------

- support for todos in api documentation
- access log improvements

0.3
----------------

- upstream server proxy (proxy an existing api)
- variables in url path
- simple api documentation (markdown)
- api authentication (HTTP Basic authentication)
- manage interface improvements

No notes on earlier releases.


mock_server/server.py
# -*- coding: utf-8 -*-

import os
import re
import datetime

import tornado.web

from data import SUPPORTED_FORMATS, SUPPORTED_MIMES, DEFAULT_FORMAT
from data import SUPPORTED_METHODS
from tornado_flash_message_mixin import FlashMessageMixin
from model import ApplicationData, ResourceMethod, RPCMethod

from methodslisting import ResourcesLoader, RPCMethodsLoader
from validators import validate_url

import model
import rest
import xmlrpc

try:
    import jsonrpc
    jsonrpc_available = True
except ImportError:
    jsonrpc_available = False


try:
    import fastrpcapi
    fastrpc_available = True
except ImportError:
    fastrpc_available = False


class Application(tornado.web.Application):
    def __init__(self, port, address, api_dir, debug, application_data):

        self.data = ApplicationData(os.path.join(api_dir, application_data))
        self.data.load()

        supported_formats = "|".join(
            map(lambda x: ".%s" % x, SUPPORTED_FORMATS.keys()))
        handlers = [
            (r"/__manage/resource/(.*)", ResourceMethodHandler),
            (r"/__manage/rpc/(.*)", RPCMethodHandler),
            (r"/__manage/upstream-server", UpstreamServerHandler),
            (r"/__manage/logs", ResourcesLogsHandler),
            (r"/__manage/create/rpc", CreateRPCMethodHandler),
            (r"/__manage/create", CreateResourceMethodHandler),
            (r"/__manage", ListResourcesHandler),
            (r"/%s" % RPCHandler.PATH, RPCHandler),
            (r"/(.*)(%s)" % supported_formats, MainHandler),
            (r"/(.*)", MainHandler),
        ]
        settings = dict(
            debug=debug,
            port=port,
            address=address,
            dir=api_dir,
            jsonrpc=jsonrpc_available,
            fastrpc=fastrpc_available,
            template_path=os.path.join(os.path.dirname(__file__), "templates"),
            static_path=os.path.join(os.path.dirname(__file__), "static"),
            static_url_prefix="/__static/",
            cookie_secret="__TODO:_GENERATE_YOUR_OWN_RANDOM_VALUE_HERE__",
        )
        super(Application, self).__init__(handlers, **settings)


class BaseHandler(tornado.web.RequestHandler):

    def set_content_type(self, content_type, charset="utf-8"):
        self.set_header("Content-Type",
                        "%s; charset=%s" % (content_type, charset))

    def set_headers(self, headers):
        if not headers:
            return
        for name, value in headers:
            self.set_header(name, value)

    def log_request(self):
        data = {
            "method": self.request.method,
            "uri": self.request.uri,
            "url_path": self.request.path,
            "status": self.get_status(),
            "remote_ip": self.request.remote_ip,
            "request_time": 1000.0 * self.request.request_time(),
            "headers": self.request.headers,
            "body": unicode(self.request.body),
            "arguments": self.request.arguments,
            "cookies": self.request.cookies,
            "time": datetime.datetime.now().isoformat()
        }

        model.add_to_resources_log(self.log_request_name, data)

    @property
    def log_request_name(self):
        return os.path.join(
            self.settings["dir"], "access-%s.log" %
            datetime.datetime.now().strftime("%Y-%m-%d"))

    def _upstream_server_callback(self, response):
        self.set_headers(response.headers)
        self.write(response.content)
        self.set_status(response.status_code)
        self.finish()


class MainHandler(BaseHandler):

    @tornado.web.asynchronous
    def head(self, path, format=DEFAULT_FORMAT):
        self._resolve_request(path, format)

    @tornado.web.asynchronous
    def get(self, path, format=DEFAULT_FORMAT):
        self._resolve_request(path, format)

    @tornado.web.asynchronous
    def post(self, path, format=DEFAULT_FORMAT):
        self._resolve_request(path, format)

    @tornado.web.asynchronous
    def delete(self, path, format=DEFAULT_FORMAT):
        self._resolve_request(path, format)

    @tornado.web.asynchronous
    def patch(self, path, format=DEFAULT_FORMAT):
        self._resolve_request(path, format)

    @tornado.web.asynchronous
    def put(self, path, format=DEFAULT_FORMAT):
        self._resolve_request(path, format)

    @tornado.web.asynchronous
    def options(self, path, format=DEFAULT_FORMAT):
        self.set_header("Access-Control-Allow-Origin", "*")
        self.set_header("Access-Control-Max-Age", 21600)

        if "Access-Control-Request-Headers" in self.request.headers:
            self.set_header(
                "Access-Control-Allow-Headers",
                self.request.headers["Access-Control-Request-Headers"])
        else:
            headers = ", ".join(x.upper() for x in self.request.headers)
            self.set_header("Access-Control-Allow-Headers", headers)

        if "Access-Control-Request-Method" in self.request.headers:
            self.set_header(
                "Access-Control-Allow-Methods",
                "OPTIONS, %s" %
                self.request.headers["Access-Control-Request-Method"])

        self.write("")
        self.finish()

    def _handle_request_on_upstream(self):
        provider = rest.UpstreamServerProvider(
            self.application.data.upstream_server)

        provider({
            "uri": self.request.uri,
            "method": self.request.method,
            "body": self.request.body,
            "headers": self.request.headers,
            "status_code": self.status_code,
            "format": self.format
        }, self._upstream_server_callback)

    def _resolve_request(self, url_path, format):
        # log request
        self.log_request()
        self.set_header("Access-Control-Allow-Origin", "*")

        # get request data
        self.format = self._get_format(format)
        method = self.request.method
        self.status_code = int(self.get_argument("__statusCode", 200))

        # upstream server
        upstream_server = self.application.data.get_upstream_server(
            "%s-%s" % (method, url_path))

        if self.application.data.upstream_server and upstream_server:
            return self._handle_request_on_upstream()

        # mock
        provider = rest.FilesMockProvider(self.settings["dir"])

        response = rest.resolve_request(
            provider, method, url_path, self.status_code, self.format)

        if provider.error:
            if self.application.data.upstream_server:
                return self._handle_request_on_upstream()

        # response
        self.set_status(response.status_code)
        self.set_content_type(SUPPORTED_FORMATS[self.format][0])
        self.set_headers(response.headers)

        self.write(response.content)
        self.finish()

    def _default_response(self, url_path, method, status_code, format):
        self.set_status(404)
        self.render("api_not_exists.html", url_path=url_path, method=method,
                    status_code=status_code, format=format)

    def _get_format(self, format):
        format = format[1:] if format[0] == "." else format

        if "Accept" in self.request.headers and\
                self.request.headers["Accept"] in SUPPORTED_MIMES:
            format = SUPPORTED_MIMES[self.request.headers["Accept"]]

        return format


class RPCHandler(BaseHandler):
    PATH = "RPC2"
    PARSE_ERROR = (-32700, "parse_error")

    @tornado.web.asynchronous
    def post(self):
        # log request
        self.log_request()
        self.set_header("Access-Control-Allow-Origin", "*")

        if not "Content-Length" in self.request.headers and \
                self.request.headers.get(
                    "Transfer-Encoding", None) == "chunked":
            self.chunks = ""
            self._stream = self.request.connection.stream
            self._stream.read_until("\r\n", self._on_chunk_length)
        else:
            self._resolve_rpclib()
            self._process(self.request.body)

    def _on_chunks(self, all_chunks):
        self._resolve_rpclib()
        self._process(all_chunks)

    def _on_chunk_length(self, data):
        assert data[-2:] == "\r\n", "chunk size ends with CRLF"
        chunk_length = int(data[:-2], 16)

        if chunk_length:
            self._stream.read_bytes(chunk_length + 2, self._on_chunk_data)
        else:
            self._on_chunks(self.chunks)

    def _on_chunk_data(self, data):
        assert data[-2:] == "\r\n", "chunk data ends with CRLF"
        self.chunks += data[:-2]

        self._stream.read_until("\r\n", self._on_chunk_length)

    def _is_jsonrpc(self, content_type):
        if self.settings["jsonrpc"] and content_type == "application/json":
            return True
        else:
            return False

    def _is_fastrpc(self, content_type):
        if self.settings["fastrpc"] and content_type == "application/x-frpc":
            return True
        else:
            return False

    def _resolve_rpclib(self):
        if "Content-Type" in self.request.headers:
            if self._is_jsonrpc(self.request.headers["Content-Type"]):
                self.content_type = "application/json"
                self.rpclib = jsonrpc
            elif self._is_fastrpc(self.request.headers["Content-Type"]):
                self.rpclib = fastrpcapi
                self.content_type = "application/x-frpc"
            else:
                self.content_type = "text/xml"
                self.rpclib = xmlrpc
        else:
            self.content_type = "text/xml"
            self.rpclib = xmlrpc

    def _handle_request_on_upstream(self, request_data):
        provider = self.rpclib.UpstreamServerProvider(
            self.application.data.upstream_server)

        if isinstance(provider, fastrpcapi.UpstreamServerProvider):
            provider(request_data, self._upstream_server_callback)
        else:
            headers = self.request.headers
            headers["Accept"] = self.content_type

            provider({
                "uri": self.request.uri,
                "method": self.request.method,
                "body": request_data,
                "headers": headers
            }, self._upstream_server_callback)

    def _process(self, request_data):
        method_name = self.rpclib.FilesMockProvider.get_method_name(
            request_data)
        self.method_name = method_name

        # upstream server
        upstream_server = self.application.data.get_rpc_upstream_server(
            method_name)

        if self.application.data.upstream_server and upstream_server:
            return self._handle_request_on_upstream(request_data)

        # mock
        provider = self.rpclib.FilesMockProvider(self.settings["dir"])
        response = provider(method_name=method_name)

        if provider.error and self.application.data.upstream_server:
            return self._handle_request_on_upstream(request_data)

        self.set_headers(response.headers)

        self.write(response.content)
        self.finish()


class ResourcesLogsHandler(BaseHandler):
    def get(self):
        self.render("resources_logs.html",
                    data=model.load_resources_log(self.log_request_name))


class ListResourcesHandler(BaseHandler, FlashMessageMixin):
    def get(self):
        # load categories
        categories, methods_with_category = \
            self.application.data.list_categories()

        # load resources
        resources_loader = ResourcesLoader(
            self.settings["dir"], self.application.data,
            SUPPORTED_METHODS)
        paths = resources_loader.load()

        # load rpc methods
        rpc_methods_loader = RPCMethodsLoader(
            self.settings["dir"], self.application.data,
            self.settings["jsonrpc"])
        rpc_methods = rpc_methods_loader.load(RPCHandler.PATH)

        # add resources to category
        for path in paths:
            for resource in path.resources.itervalues():
                if resource.id in methods_with_category:
                    category = methods_with_category[resource.id]["category"]
                else:
                    category = "__default"

                if path not in categories[category]:
                    categories[category].append(("rest", path))
                    break

        # add rpc methods to category
        for method in rpc_methods:
            name = "RPC-%s" % method["name"]
            if name in methods_with_category:
                category = methods_with_category[name]["category"]
            else:
                category = "__default"

            categories[category].append(("rpc", method))

        # render to response
        self.render("list_resources.html", categories=categories,
                    application_data=self.application.data,
                    number_of_methods=len(methods_with_category),
                    some_method=bool(paths or rpc_methods),
                    flash_messages=self.get_flash_messages(
                        ("success", "error")))


class CreateResourceMethodHandler(BaseHandler, FlashMessageMixin):

    def get(self):
        url_path = self.get_argument("url_path", "")
        method = self.get_argument("method", None)

        method_file = ResourceMethod(
            self.settings["dir"], url_path, method)
        method_file.load_responses()
        method_file.load_description()

        if method is None:
            category = ""
        else:
            category = self.application.data.get_category(method_file.id)

        self.render(
            "create_resource.html",
            protocol="rest", category=category, method_file=method_file,
            SUPPORTED_FORMATS=SUPPORTED_FORMATS.keys(),
            jsonrpc=self.settings["jsonrpc"])

    def post(self):
        # check xsrf cookie
        self.check_xsrf_cookie()

        # get data from request body
        data = tornado.escape.json_decode(self.request.body)

        # save resource
        for response in data["responses"]:
            # save resource
            method_file = ResourceMethod(
                self.settings["dir"], data["url_path"], data["method"])
            method_file.description = data["description"]
            method_file.add_response(
                response["status_code"], response["format"],
                response["body"], response["headers"])
            method_file.save()

            # add resource to category
            self.application.data.save_category(
                method_file.id, data["category"])

        self.set_flash_message(
            "success",
            "Resource '%s' has been successfully created." % data["url_path"])
        self.set_header("Content-Type", "application/json")
        self.write("OK")


class CreateRPCMethodHandler(BaseHandler, FlashMessageMixin):
    def get(self):
        method_name = self.get_argument("method_name", "")

        method_file = RPCMethod(
            os.path.join(self.settings["dir"], RPCHandler.PATH))
        method_file.load(method_name)
        method_file.load_description()

        category = self.application.data.get_rpc_category(method_name)

        self.render(
            "create_resource.html",
            protocol="rpc", category=category, method_file=method_file,
            SUPPORTED_FORMATS=SUPPORTED_FORMATS.keys(),
            jsonrpc=self.settings["jsonrpc"])

    def post(self):

        self.check_xsrf_cookie()

        method_name = self.get_argument("method_name")
        method_response = self.get_argument("method_response")
        category = self.get_argument("category", "")
        description = self.get_argument("description", "")

        # save method
        rpc_file = RPCMethod(
            os.path.join(self.settings["dir"], RPCHandler.PATH))
        rpc_file.description = description
        rpc_file.save(method_name, method_response)

        # add method to category
        self.application.data.save_category(
            "RPC-%s" % method_name, category)

        self.set_flash_message(
            "success",
            "RPC method '%s' has been successfully created." % method_name)
        self.redirect("/__manage")


class UpstreamServerHandler(BaseHandler, FlashMessageMixin):
    def post(self):
        self.check_xsrf_cookie()

        upstream_server = self.get_argument("upstream_server", "")

        if upstream_server and not validate_url(upstream_server):
            self.set_flash_message(
                "error",
                "Url is not valid.")
            return self.redirect("/__manage")

        self.application.data.upstream_server = upstream_server
        self.application.data.save()

        self.set_flash_message(
            "success",
            "Upstream server has been successfully saved.")
        self.redirect("/__manage")


class ResourceMethodHandler(BaseHandler, FlashMessageMixin):
    def get(self, resource):
        _method = self.get_argument("_method", "get")
        if _method == "delete":
            return self.delete(resource)

        raise tornado.web.HTTPError(405)

    def post(self, resource):
        self.check_xsrf_cookie()

        upstream_server = bool(self.get_argument("upstream_server", None))

        self.application.data.save_method_upstream_server(
            resource, upstream_server)

        self.set_flash_message(
            "success",
            "Upstream server for resource method has been successfully %s." %
            ("active" if upstream_server else "deactivate"))

        self.redirect("/__manage")

    def delete(self, resource):
        # get http method, status and url path
        c = re.compile(r"(%s)-(.*)" %
                       ("|".join(SUPPORTED_METHODS)))
        m = c.match(resource)
        method, url_path = m.groups()

        # delete resource method
        method_file = ResourceMethod(self.settings["dir"], url_path, method)
        method_file.delete()

        self.application.data.delete_resource(resource)

        # redirect
        self.set_flash_message(
            "success",
            "Resource method has been successfully deleted.")

        self.redirect("/__manage")


class RPCMethodHandler(BaseHandler, FlashMessageMixin):
    def get(self, method_name):
        _method = self.get_argument("_method", "get")
        if _method == "delete":
            return self.delete(method_name)

        raise tornado.web.HTTPError(405)

    def post(self, method_name):
        self.check_xsrf_cookie()

        upstream_server = bool(self.get_argument("upstream_server", None))

        self.application.data.save_method_upstream_server(
            "RPC-%s" % method_name, upstream_server)

        self.set_flash_message(
            "success",
            "Upstream server for RPC method has been successfully %s." %
            ("active" if upstream_server else "deactivate"))

        self.redirect("/__manage")

    def delete(self, method_name):
        rpc_file = RPCMethod(
            os.path.join(self.settings["dir"], RPCHandler.PATH))
        rpc_file.delete(method_name)

        self.application.data.delete_resource("RPC-%s" % method_name)

        self.set_flash_message(
            "success",
            "RPC method has been successfully deleted.")

        self.redirect("/__manage")
